### ISSUE TITLE
Picture element inside polymer-elment doesn't get container width

### ISSUE THREAD

**Issue by scottnath** [scottjehl/picturefill#412]
Hi there,

I'm trying to use the <picture> element inside of a web component I'm creating, but it is choosing the smallest possible image as `<picture>` is not getting any dimensions from the container element.

I have tried giving :host a display:block among a myriad of other atempts at doing this with styling, including making sure the container is 100%, but nothing works. 

Here is a [codepen of what I'm trying to do](http://codepen.io/scottnath/pen/RNoBXd) showing the polymer-made element using the smallest image, then the `<picture>` element on it's own showing the correct image.

Note in the live codepen version that the `<hr>` (horizontal rule) elements created by the polymer-element are the full width and thus understand the correct width of the container.

This is the code in the codepen:

```
<script src="http://www.polymer-project.org/platform.js"></script>
<script src="http://www.polymer-project.org/polymer.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/picturefill/2.2.0/picturefill.min.js"></script>

<!-- Define the element here -->
<polymer-element name="picture-container" attributes="value">
  <template>
    <style>
      :host {
      display: block;
        border: 1px solid #000;
        padding: 1em;
      }
    </style>
     
    <hr>
      <h2>Picture In Polymer Element</h2>
      <picture>
        <source srcset="http://www.placecage.com/c/500/500" media="(min-width: 500px)">
        <source srcset="http://www.placecage.com/c/200/200" media="(min-width: 400px)">
        <img srcset="http://www.placecage.com/c/100/100" alt="An example of the picture element">
      </picture>
    <hr>
    
  </template>
  <script>
    Polymer('picture-container');
  </script>
</polymer-element>

<picture-container></picture-container>

<hr>
  <h2>Picture In Polymer Element</h2>
  <picture>
    <source srcset="http://www.placecage.com/c/500/500" media="(min-width: 500px)">
    <source srcset="http://www.placecage.com/c/200/200" media="(min-width: 400px)">
    <img srcset="http://www.placecage.com/c/100/100" alt="An example of the picture element">
  </picture>
<hr>
```

any help on this greatly appreciated.

thanks,
Scott

**Comment by mike-engel** [comment]
@scottnath Is there a way to detect when a polymer element has been added to the DOM? If so, can you run `picturefill()` after it's been added? I have a feeling, and may be wrong, that picturefill isn't finding the right image because it's not in the DOM yet. Also, what browser is this running on? Does it work on Chrome but not firefox or vice-vera?

**Comment by scottnath** [comment]
@mike-engel Good idea. I tried it, but it made no change. There is a [`ready()` method](https://www.polymer-project.org/docs/polymer/polymer.html#global) for Polymer which fires after the element is ready.

I changed my [codepen](http://codepen.io/scottnath/pen/RNoBXd) to include running `picturefill()` and it looked like this:

```
...
  </template>
  <script>
    Polymer('picture-container',{
      ready: function() {
        picturefill();
      }
    });
  </script>
</polymer-element>
...
```

that made no changes I could determine and the in-element picture is still the wrong size.

I have only tested this on chrome - where it should work fine.

**Comment by scottnath** [comment]
Hi @aFarkas 

I tried this, but it didn't help. You can see the updated codepen here:
http://codepen.io/scottnath/pen/Wboayg
which contains your code fix. I also tried this idea on my local environment and it didn't help.

For your q...for the larger project I'm working on, I do need a web component. This very minor POC is to solve one aspect of the project I'm building - which requires me to find a web-component-related fix.

thanks,
Scott

**Comment by aFarkas** [comment]
For me it is working now in FF, before it didn't work in FF at all.

**Comment by scottnath** [comment]
Aha! Excellent. Yes the first version: http://codepen.io/scottnath/pen/RNoBXd does not work in firefox, but the version with your code: http://codepen.io/scottnath/pen/Wboayg works excellent. It does not, however, work in Chrome. It does work in Safari though.

Can you explain what you meant by "*Chrome: I think I know what's happening the media from another window than you window is used. But don't know whether this is a bug or intended.*"? Chrome thinks that the element-internal `<picture>` is from a different window?

thanks,
Scott

**Comment by aFarkas** [comment]
Here is a testcase: http://jsfiddle.net/5td6ym5L/embedded/result/

**Comment by scottnath** [comment]
Thanks for the testcase. I'm not sure what lesson I can take away from it though.

## firefox
boxes and iframe don't show up at all

## safari
both boxes are 600

## chrome & chrome canary
outside iframe: 1000px
inside: 101px

AFAIK, `template` is not an iframe and the picture element is supported by chrome.

Regardless...it seems that chrome is the culprit here - but does this mean that chrome's implementation of picture is broken somehow?

## other firefox issues
looking at my codepen with @aFarkas 's codefix in it:
http://codepen.io/scottnath/pen/Wboayg
I discovered that in Firefox (33.03) if I go into `about:config` and change these two settings to true:
```
dom.image.srcset.enabled
dom.image.picture.enabled
```
then this mimics the incorrect behavior of chrome for the picture element. However, the @aFarkas testcase still does not work.

**Comment by aFarkas** [comment]
@scottnath 
The testcase is only build for Chrome. (native supporting browsers)

But your test with firefox flags on is indeed an important one. I will write an extended testcase (or if you scream *here*, you can build those for us):

a) iframes
b) template
c) document fragments
d) shadowdom

But you have to know. Nothing of this is a picturefill related bug.

**Comment by scottnath** [comment]
So...do you know what this bug is then? I'm not sure how to search for it to find a solution...

**Comment by aFarkas** [comment]
It's nothing you can change. But you can open a ticket. Here is the testcase using the template element: http://jsfiddle.net/5td6ym5L/2/embedded/result/

The other things seem to work.

**Comment by aFarkas** [comment]
The corresponding [Chrome bug was fixed](https://code.google.com/p/chromium/issues/detail?id=446876).

**Comment by scottnath** [comment]
@aFarkas great job! Seeing the fix in my Chrome Canary already. Thanks for submitting that!
-Scott

**Comment by nazar-pc** [comment]
I found a bit more convenient way.
Since we can extend native elements - I've created custom elements based on native `picture` and `img`, now I'm using them as:
```html
<img is="cs-picturefill-img" srcset="whatever.png">
```
Similarly for `picture`.
Under the hood it just calls `picturefill()` function on itself when element is created.
[Sources of both](https://github.com/nazar-pc/CleverStyle-CMS/tree/master/components/plugins/Picturefill/includes/html)
